runtime/v1/linux/runtime.go

```diff
@@ -191,23 +191,24 @@ func (r *Runtime) Create(ctx context.Context, id string, opts runtime.CreateOpts
 		}
 		exitHandler := func() {
 			log.G(ctx).WithField("id", id).Info("shim reaped")
+
 			t, err := r.tasks.Get(ctx, id)
 			if err != nil {
 				// Task was never started or was already successfully deleted
 				return
 			}
 			lc := t.(*Task)
 
-			log.G(ctx).WithFields(logrus.Fields{
-				"id":        id,
-				"namespace": namespace,
-			}).Warn("cleaning up after killed shim")
-			if err = r.cleanupAfterDeadShim(context.Background(), bundle, namespace, id, lc.pid); err != nil {
+			if err = r.cleanupAfterDeadShim(context.Background(), bundle, namespace, id); err != nil {
 				log.G(ctx).WithError(err).WithFields(logrus.Fields{
 					"id":        id,
 					"namespace": namespace,
 				}).Warn("failed to clean up after killed shim")
 			}
+
+			if err := lc.shim.KillShim(ctx); err != nil {
```

Comment:
The `exitHandler` will only be called when the shim process exits, there is no need to kill it again.

I do agree that the `cleanupAfterDeadShim` onClose function installed by `loadTasks` may try killing the shim. But please not that for shim v2 we can no longer do that, because shim v2 implementation may have per-pod shim process, which can't be killed when one container dies.

Reply:
I get some time to check the code, and try to explain my logic:
1.  in exit handler, the normal case should be shim exit and resources be cleanup, but in bad case, still some shim hangover. so I want to add one more kill to shim
2. in loadTasks, I do not found the good place to do killShim, we get `shimclient` through `bundle.NewShimClient`, but if not get response, we can not do `KillShim`